### PR TITLE
Windows: Hint to run client elevated

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

@thaJeztah. I've seen countless reports now where users aren't reading the quick start documentation for Docker on Windows carefully enough, and are confused by the message they get back using the CLI.

eg in French:

image

or in English:

image

So this PR does a little to hopefully address it by adding a hint to the error message to run elevated when using the default configuration named pipe //./pipe/docker_engine.

And the new message:

image

@friism @dmp42 @mavenugo This could be a useful one for consideration.

### CODE REVIEW

**Comment by thaJeztah** [comment]
Should we use the same / similar error as was recently added for Linux? https://github.com/docker/docker/pull/26797/files

Can we distinguish between "access denied" and "not running" on Windows?

**Comment by jhowardmsft** [comment]
Ah we can't distinguish.

**Comment by jhowardmsft** [comment]
@thaJeztah You OK with this? It's come up a few times over the past few days too....

File: client/request.go [hunk 1]
```diff
@@ -143,6 +143,20 @@ func (cli *Client) sendClientRequest(ctx context.Context, method, path string, q
...
+		// daemon on Windows where the daemon is listening on a named pipe
+		// `//./pipe/docker_engine, and the client must be running elevated.
+		// Give users a clue rather than the not-overly useful message
+		// such as `error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.25/info:
+		// open //./pipe/docker_engine: The system cannot find the file specified.`.
+		// Note we can't string compare "The system cannot find the file specified" as
+		// this is localised - for example in French the error would be
+		// `open //./pipe/docker_engine: Le fichier spécifié est introuvable.`
+		if strings.Contains(err.Error(), `open //./pipe/docker_engine`) {
+			err = errors.New(err.Error() + " In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running.")
```
Comment (thaJeztah) [inline client/request.go]: "In the default daemon configuration on Windows" is probably a bit too verbose; Would this work?

    Make sure the docker daemon is running, and the client is running in elevated mode.

Comment (thaJeztah) [inline client/request.go]: I understand, but that's with any setup, I don't think we can take every possibility into account. IIUC, even with a modified setup, Reasons why it doesn't work is;

- daemon not running
- client doesn't have the required permissions

Correct?

Comment (jhowardmsft) [inline client/request.go]: Yes, these are the big two I see every day. But always also in the default configuration just following the guides.

**Comment by jhowardmsft** [comment]
@vdemeester

File: client/request.go (hunk 1 above)
Comment (mstanleyjones) [inline client/request.go]: How about something like "Either the Docker daemon is not running, it is not elevated, or there is another problem with the configuration of the Docker daemon."

**Review by mdlinville — CHANGES_REQUESTED** [review]

File: client/request.go (hunk 1 above)
Comment (jhowardmsft) [inline client/request.go]: Ah no. Not quite. The #1 reason is the the client is not running elevated. Most likely the daemon is running, but we can't distinguish between lack of elevation of the client and the daemon not running. The daemon not running is very much a secondary issue, but has come up occasionally.

File: client/request.go [hunk 2]
```diff
@@ -143,6 +143,20 @@ func (cli *Client) sendClientRequest(ctx context.Context, method, path string, q
...
+		// Although there's not a strongly typed error for this in go-winio,
+		// lots of people are using the default configuration for the docker
+		// daemon on Windows where the daemon is listening on a named pipe
+		// `//./pipe/docker_engine, and the client must be running elevated.
+		// Give users a clue rather than the not-overly useful message
+		// such as `error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.25/info:
+		// open //./pipe/docker_engine: The system cannot find the file specified.`.
+		// Note we can't string compare "The system cannot find the file specified" as
+		// this is localised - for example in French the error would be
+		// `open //./pipe/docker_engine: Le fichier spécifié est introuvable.`
```
Comment (vdemeester) [inline client/request.go]: :heart:

File: client/request.go (hunk 1 above)
Comment (vdemeester) [inline client/request.go]: @jhowardmsft The client must be elevated only if it's the default setup (or at least if the daemon and client discuss through a named pipe), am I right ? 

If yes, it's a little bit more than *default daemon configuration* ; maybe we could add something about "*if the daemon is listening via named pipe*" or something ?

**Review by vdemeester — COMMENTED** [review]
2 comments (one completely useless :stuck_out_tongue_closed_eyes:)

File: client/request.go (hunk 2 above)
Comment (jhowardmsft) [inline client/request.go]: :France:

**Comment by jhowardmsft** [comment]
Elevation is required only using names pipe. And only if (as per default) the pipe configuration hasn't been changed such that other Windows groups can access the pipe. By default it's administrators only. The problem I have with saying named pipe explicitly is that I'm not sure it's going to mean much to users starting out with docker.  We only know that the client is attempting to connect to a local named pipe with this change. If remote, I'm not so concerned as the user would likely already know about -H and daemon configuration.

**Comment by vdemeester** [comment]
@jhowardmsft I see :sweat_smile: 
LGTM for me then 👼

**Comment by jhowardmsft** [comment]
@mstanleyjones PTAL at the above comments.

**Comment by mstanleyjones** [comment]
I don't have a strong preference, but I'm not sure if the message actually will alleviate confusion. @friism WDYT?

**Comment by friism** [comment]
What about "Check that the Docker daemon is up and that the client is running as an elevated process."

If we document this with more detail in the FAQ, hopefully people will find it when googling.

**Comment by jhowardmsft** [comment]
Yes, but... it would be giving incorrect information(to elevate) potentially unnecessarily.

**Comment by crosbymichael** [comment]
LGTM

Much better error message than what was originally returned to the user.

**Pull request merged**